### ISSUE TITLE
Reused connection not reauthorising when auth details change

### ISSUE THREAD

**Issue by paulharris** [curl/curl#1685]
Hi again,

This one is interesting,
I believe I have the same curl library on both Windows and Linux.
I'm compiling both with CMake, and I have ticked the same options on both.  I don't think it matters much which options, but I can't be sure yet.

The code is derived from "curl --libcurl", I've just executed the same query twice on a URL, with different user/pass (for the digest auth in this case).

For Linux, it works as expected.
For Windows, the second connection authenticates with the FIRST username.

The server has a simple user/pass setup, pointing to a simple CGI that prints out the auth username to a log file.
You could probably do the same with a static webpage, and webserver logs of what user authenticates with the page.

```
#include <curl/curl.h>

int main(int argc, char *argv[])
{
  CURLcode ret;
  CURL *hnd;

#define URL "http://localhost/myurl"

  hnd = curl_easy_init();
  curl_easy_setopt(hnd, CURLOPT_BUFFERSIZE, 102400L);
  curl_easy_setopt(hnd, CURLOPT_URL, URL);
  curl_easy_setopt(hnd, CURLOPT_NOPROGRESS, 1L);
  curl_easy_setopt(hnd, CURLOPT_USERPWD, "user_primary:user_primary");
  curl_easy_setopt(hnd, CURLOPT_USERAGENT, "curl/7.55.0-DEV");
  curl_easy_setopt(hnd, CURLOPT_MAXREDIRS, 50L);
  curl_easy_setopt(hnd, CURLOPT_HTTPAUTH, (long)CURLAUTH_DIGEST);
  curl_easy_setopt(hnd, CURLOPT_SSH_KNOWNHOSTS, "/home/paul/.ssh/known_hosts");
  curl_easy_setopt(hnd, CURLOPT_TCP_KEEPALIVE, 1L);

  ret = curl_easy_perform(hnd);


  // set to 1 to show the bug
#define SHARE_CONNS 1

#if SHARE_CONNS
  curl_easy_reset(hnd);
#else
  curl_easy_cleanup(hnd);
  hnd = curl_easy_init();
#endif

  curl_easy_setopt(hnd, CURLOPT_BUFFERSIZE, 102400L);
  curl_easy_setopt(hnd, CURLOPT_URL, URL);
  curl_easy_setopt(hnd, CURLOPT_NOPROGRESS, 1L);
  curl_easy_setopt(hnd, CURLOPT_USERPWD, "user_replica:user_replica");
  curl_easy_setopt(hnd, CURLOPT_USERAGENT, "curl/7.55.0-DEV");
  curl_easy_setopt(hnd, CURLOPT_MAXREDIRS, 50L);
  curl_easy_setopt(hnd, CURLOPT_HTTPAUTH, (long)CURLAUTH_DIGEST);
  curl_easy_setopt(hnd, CURLOPT_SSH_KNOWNHOSTS, "/home/paul/.ssh/known_hosts");
  curl_easy_setopt(hnd, CURLOPT_TCP_KEEPALIVE, 1L);

  ret = curl_easy_perform(hnd);

  curl_easy_cleanup(hnd);

  hnd = NULL;

  return (int)ret;
}

```

Thoughts?
cheers,
Paul

**Comment by jay** [comment]
Please give the curl_version()s in your report. I can reproduce this in SSPI-enabled versions. There are two sets of digest code and they differ somewhat. Both take credentials and use that to create an authorization but SSPI-enabled builds [may update a previously generated context](https://github.com/curl/curl/blob/curl-7_54_1/lib/vauth/digest_sspi.c#L441-L470) for the host. The [digest data](https://github.com/curl/curl/blob/curl-7_54_1/lib/urldata.h#L414-L431) includes that context and is not cleared [during easy reset](https://github.com/curl/curl/blob/curl-7_54_1/lib/easy.c#L990). Is there some reason we can't call Curl_digest_cleanup there? This or something similar has come up before I think, if we reset it then the next request doesn't send the auth header (resulting in another 401)

~~~diff
diff --git a/lib/easy.c b/lib/easy.c
index 2b1ce9e..ee0c9d8 100644
--- a/lib/easy.c
+++ b/lib/easy.c
@@ -72,6 +72,7 @@
 #include "multiif.h"
 #include "sigpipe.h"
 #include "ssh.h"
+#include "http_digest.h"
 /* The last 3 #include files should be in this order */
 #include "curl_printf.h"
 #include "curl_memory.h"
@@ -990,6 +991,7 @@ void curl_easy_reset(struct Curl_easy *data)
   /* zero out authentication data: */
   memset(&data->state.authhost, 0, sizeof(struct auth));
   memset(&data->state.authproxy, 0, sizeof(struct auth));
+  Curl_digest_cleanup(data);
 }
 
 /*
~~~

**Comment by paulharris** [comment]
I should've mentioned, it was the current git-master copy, whatever the
version is at the moment...

**Comment by frenche** [comment]
@jay, but I'd expect the user to get updated in the second request even if easy_reset was not called, no?
If we do some caching of the sspi identity, then perhaps we should userpwd in digestdata and compare before reusing cached identity.

**Comment by jay** [comment]
> I'd expect the user to get updated in the second request even if easy_reset was not called, no?

right you are, please review #1742

**Comment by jay** [comment]
Fixed in 0b5665c, thanks guys
